src/cargo/core/resolver/mod.rs

```diff
@@ -247,158 +193,245 @@ pub fn resolve(summary: &Summary, method: &Method,
-    //    package
-    // 2. There are no activated versions for this package which are
-    //    semver-compatible, or there's an activated version which is
-    //    precisely equal to `candidate`.
-    //
-    // This means that we're going to attempt to activate each candidate in
-    // turn. We could possibly fail to activate each candidate, so we try
-    // each one in turn.
-    let mut last_err = None;
-    for candidate in my_candidates {
         trace!("{}[{}]>{} trying {}", parent.name(), cur, dep.name(),
                candidate.version());
-        let mut my_cx = cx.clone();
-        my_cx.resolve.graph.link(parent.package_id().clone(),
-                                 candidate.package_id().clone());
+        cx.resolve.graph.link(parent.package_id().clone(),
+                              candidate.package_id().clone());
 
         // If we hit an intransitive dependency then clear out the visitation
         // list as we can't induce a cycle through transitive dependencies.
         if !dep.is_transitive() {
-            my_cx.visited.clear();
-        }
-        let my_cx = try!(activate(my_cx, registry, candidate, &method,
-                                  &mut |cx, registry| {
-            activate_deps(cx, registry, parent, deps.clone(), cur + 1, finished)
-        }));
-        match my_cx {
-            Ok(cx) => return Ok(Ok(cx)),
-            Err(e) => { last_err = Some(e); }
+            cx.visited.clear();
         }
+        remaining_deps.extend(try!(activate(&mut cx, registry,
+                                            candidate, &method)));
     }
-    trace!("{}[{}]>{} -- {:?}", parent.name(), cur, dep.name(), last_err);
+    trace!("resolved: {:?}", cx.resolve);
+    Ok(cx.resolve)
+}
 
-    // Oh well, we couldn't activate any of the candidates, so we just can't
-    // activate this dependency at all
-    Ok(activation_error(&cx, registry, last_err, parent, dep, prev_active,
-                        &candidates))
+// Searches up `backtrack_stack` until it finds a dependency with remaining
+// candidates. Resets `cx` and `remaining_deps` to that level and returns the
+// next candidate. If all candidates have been exhausted, returns an activation
+// error.
+fn find_candidate(backtrack_stack: &mut Vec<BacktrackFrame>,
+                  cx: &mut Context, remaining_deps: &mut Vec<DepsFrame>,
+                  parent: &mut Rc<Summary>, cur: &mut usize,
+                  dep: &mut Rc<Dependency>,
+                  err: Box<CargoError>) -> CargoResult<Rc<Summary>> {
+    while let Some(mut frame) = backtrack_stack.pop() {
+        match frame.remaining_candidates.next() {
+            None => {
+                let top_deps_frame = frame.deps_backup.last().unwrap();
+                trace!("{}[{}]>{} -- {:?}", frame.parent.name(),
+                       top_deps_frame.remaining_siblings.cur_index(),
+                       frame.dep.name(), Some(&err));
```

Comment:
It's not necessarily super important that the logging here remains exactly the same, for example it may not be worth it to bend over backwards to get the same information. Along those lines I think it's fine to shorten this or just elide it entirely. These sorts of tracing messages can always be added in an on-demand fashion later.